File: php/tests/WebhookTest.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Tests;

use Fivexer\SDK\Exception\FivexerException;
use Fivexer\SDK\Webhook\Webhook;
use PHPUnit\Framework\TestCase;

/**
 * Webhook signature verification: replay window, malformed headers, wrong secret,
 * tampered body, custom tolerance, and the platform header constant.
 */
final class WebhookTest extends TestCase
{
    private const SECRET = 'whsec_testsecret';
    private const NOW = 1750000000000;

    private static function body(): string
    {
        return '{"event":"task.matched","taskId":"task_8fk2","workerId":"agent_1"}';
    }

    private static function sign(string $secret, int $timestamp, string $body): string
    {
        return 't=' . $timestamp . ',v1=' . \hash_hmac('sha256', $timestamp . '.' . $body, $secret);
    }

    public function test_a_validly_signed_webhook_is_verified_and_parsed(): void
    {
        $raw = self::body();
        $header = self::sign(self::SECRET, self::NOW, $raw);

        $event = Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);

        self::assertSame('task.matched', $event->event);
        self::assertSame('task_8fk2', $event->data['taskId']);
        self::assertSame('agent_1', $event->data['workerId']);
    }

    public function test_a_webhook_signed_with_the_wrong_secret_is_rejected(): void
    {
        $raw = self::body();
        $header = self::sign('whsec_wrong', self::NOW, $raw);

        $this->expectException(FivexerException::class);
        Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);
    }

    public function test_a_webhook_outside_the_replay_window_is_rejected(): void
    {
        $raw = self::body();
        $header = self::sign(self::SECRET, self::NOW - (10 * 60 * 1000), $raw); // 10 min ago

        $this->expectException(FivexerException::class);
        Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);
    }

    public function test_a_tampered_body_fails_verification(): void
    {
        $raw = self::body();
        $header = self::sign(self::SECRET, self::NOW, $raw);
        $tampered = \str_replace('task_8fk2', 'task_9999', $raw);

        $this->expectException(FivexerException::class);
        Webhook::constructEvent($tampered, $header, self::SECRET, nowMs: self::NOW);
    }

    public function test_a_custom_tolerance_window_can_be_widened(): void
    {
        $raw = self::body();
        $header = self::sign(self::SECRET, self::NOW - (6 * 60 * 1000), $raw); // 6 min old

        // default tolerance (5 min) rejects
        try {
            Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);
            self::fail('expected rejection with default tolerance');
        } catch (FivexerException) {
        }
        // widened tolerance (10 min) accepts
        $event = Webhook::constructEvent($raw, $header, self::SECRET, toleranceMs: 10 * 60 * 1000, nowMs: self::NOW);
        self::assertSame('task.matched', $event->event);
    }

    public function test_malformed_signature_headers_are_rejected(): void
    {
        $raw = self::body();
        $malformed = [
            '',
            'garbage',
            't=abc,v1=def',
            't=123',
            'v1=abc',
            't=123,v1=' . \str_repeat('x', 64),
            'x=1,v1=' . \str_repeat('0', 64),
        ];
        foreach ($malformed as $bad) {
            self::assertFalse(
                Webhook::verifySignature($raw, $bad, self::SECRET, nowMs: self::NOW),
                "expected rejection for header: {$bad}",
            );
        }
    }

    public function test_a_null_webhook_header_is_rejected(): void
    {
        self::assertFalse(Webhook::verifySignature('{}', '', self::SECRET, nowMs: self::NOW));
    }

    public function test_a_webhook_with_correct_t_part_but_wrong_v_prefix_is_rejected(): void
    {
        $header = 't=' . self::NOW . ',v2=' . \str_repeat('0', 64);
        self::assertFalse(Webhook::verifySignature('{}', $header, self::SECRET, nowMs: self::NOW));
    }

    public function test_a_webhook_with_an_empty_timestamp_is_rejected(): void
    {
        $header = 't=,v1=' . \str_repeat('0', 64);
        self::assertFalse(Webhook::verifySignature('{}', $header, self::SECRET, nowMs: self::NOW));
    }

    public function test_a_payload_without_an_event_key_yields_null_event(): void
    {
        $raw = '{"taskId":"task_1","workerId":"a1"}';
        $header = self::sign(self::SECRET, self::NOW, $raw);

        $event = Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);
        self::assertNull($event->event);
        self::assertSame('task_1', $event->data['taskId']);
    }

    public function test_a_webhook_payload_with_a_non_primitive_event_yields_null_event(): void
    {
        $raw = '{"event":[1,2]}';
        $header = self::sign(self::SECRET, self::NOW, $raw);

        $event = Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);
        self::assertNull($event->event);
        self::assertNotNull($event->data);
    }

    public function test_a_non_object_webhook_payload_yields_null_event_and_data(): void
    {
        $raw = '[1,2,3]';
        $header = self::sign(self::SECRET, self::NOW, $raw);

        $event = Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);
        self::assertNull($event->event);
        self::assertNull($event->data);
    }

    public function test_a_non_json_webhook_payload_is_rejected_with_a_clear_error(): void
    {
        $raw = 'not json';
        $header = self::sign(self::SECRET, self::NOW, $raw);

        // signature passes, then parsing must fail clearly
        $this->expectException(FivexerException::class);
        Webhook::constructEvent($raw, $header, self::SECRET, nowMs: self::NOW);
    }

    public function test_signature_header_constant_is_the_platform_value(): void
    {
        self::assertSame('x-fivexer-signature', Webhook::SIGNATURE_HEADER);
    }

    public function test_verify_signature_returns_false_without_raising_for_garbage(): void
    {
        self::assertFalse(Webhook::verifySignature('{}', 'garbage', self::SECRET, nowMs: self::NOW));
    }
}

File: java/src/test/java/io/fivexer/sdk/WebhookTest.java
package io.fivexer.sdk;

import io.fivexer.sdk.webhook.Webhook;
import io.fivexer.sdk.webhook.WebhookEvent;
import org.junit.jupiter.api.Test;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Webhook signature verification: replay window, malformed headers, wrong secret,
 * tampered body, custom tolerance, and the platform header constant.
 */
class WebhookTest {

    private static final String SECRET = "whsec_testsecret";
    private static final long NOW = 1750000000000L;

    private static byte[] body() {
        return "{\"event\":\"task.matched\",\"taskId\":\"task_8fk2\",\"workerId\":\"agent_1\"}"
                .getBytes(StandardCharsets.UTF_8);
    }

    private static String sign(String secret, long timestamp, byte[] body) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
        mac.update(Long.toString(timestamp).getBytes(StandardCharsets.US_ASCII));
        mac.update((byte) '.');
        mac.update(body);
        return "t=" + timestamp + ",v1=" + bytesToHex(mac.doFinal());
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    @Test
    void a_validly_signed_webhook_is_verified_and_parsed() throws Exception {
        byte[] raw = body();
        String header = sign(SECRET, NOW, raw);

        WebhookEvent event = Webhook.constructEvent(raw, header, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW);

        assertEquals("task.matched", event.getEvent());
        assertEquals("task_8fk2", event.getData().get("taskId"));
        assertEquals("agent_1", event.getData().get("workerId"));
    }

    @Test
    void a_webhook_signed_with_the_wrong_secret_is_rejected() throws Exception {
        byte[] raw = body();
        String header = sign("whsec_wrong", NOW, raw);

        assertThrows(FivexerException.class,
                () -> Webhook.constructEvent(raw, header, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW));
    }

    @Test
    void a_webhook_outside_the_replay_window_is_rejected() throws Exception {
        byte[] raw = body();
        String header = sign(SECRET, NOW - (10L * 60 * 1000), raw); // 10 min ago

        assertThrows(FivexerException.class,
                () -> Webhook.constructEvent(raw, header, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW));
    }

    @Test
    void a_tampered_body_fails_verification() throws Exception {
        byte[] raw = body();
        String header = sign(SECRET, NOW, raw);
        byte[] tampered = new String(raw, StandardCharsets.UTF_8).replace("task_8fk2", "task_9999")
                .getBytes(StandardCharsets.UTF_8);

        assertThrows(FivexerException.class,
                () -> Webhook.constructEvent(tampered, header, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW));
    }

    @Test
    void a_custom_tolerance_window_can_be_widened() throws Exception {
        byte[] raw = body();
        String header = sign(SECRET, NOW - (6L * 60 * 1000), raw); // 6 min old

        // default tolerance (5 min) rejects
        assertThrows(FivexerException.class,
                () -> Webhook.constructEvent(raw, header, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW));
        // widened tolerance (10 min) accepts
        WebhookEvent event = Webhook.constructEvent(raw, header, SECRET, 10L * 60 * 1000, NOW);
        assertEquals("task.matched", event.getEvent());
    }

    @Test
    void malformed_signature_headers_are_rejected() {
        byte[] raw = body();
        for (String bad : new String[]{"", "garbage", "t=abc,v1=def", "t=123", "v1=abc",
                "t=123,v1=" + "x".repeat(64), "x=1,v1=" + "0".repeat(64)}) {
            assertThrows(FivexerException.class,
                    () -> Webhook.constructEvent(raw, bad, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW),
                    "expected rejection for header: " + bad);
        }
    }

    @Test
    void a_payload_without_an_event_key_yields_null_event() throws Exception {
        byte[] raw = "{\"taskId\":\"task_1\",\"workerId\":\"a1\"}".getBytes(StandardCharsets.UTF_8);
        String header = sign(SECRET, NOW, raw);

        WebhookEvent event = Webhook.constructEvent(raw, header, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW);
        assertNull(event.getEvent());
        assertEquals("task_1", event.getData().get("taskId"));
    }

    @Test
    void signature_header_constant_is_the_platform_value() {
        assertEquals("x-fivexer-signature", Webhook.SIGNATURE_HEADER);
    }

    @Test
    void verify_signature_returns_false_without_raising_for_garbage() {
        assertFalse(Webhook.verifySignature(new byte[]{}, "garbage", SECRET,
                Webhook.DEFAULT_TOLERANCE_MS, NOW));
    }

    @Test
    void non_json_payload_is_rejected_with_a_clear_error() {
        byte[] raw = "not json".getBytes(StandardCharsets.UTF_8);
        // sign it so the signature passes, then parsing must fail clearly
        assertThrows(FivexerException.class, () -> {
            String header = sign(SECRET, NOW, raw);
            Webhook.constructEvent(raw, header, SECRET, Webhook.DEFAULT_TOLERANCE_MS, NOW);
        });
    }
}
